Retouch GUI

edgardoh opened the pull request:
This are GUI changes from PR#1548.

-add an option to allow skip mouse events while editing masks
can be found on preferences->shortcuts->views->darkroom->allow to pan & zoom while editing masks

-change resize step and minimum size of brush

-resize don't care where the mouse is inside a shape

-add a preview when creating a circle

-resize a path even if on a node or segment

TurboGit commented:
> -add a preview when creating a circle

would be nice to do the same for ellipse I would say. this preview is helpful.

edgardoh commented:
Yes, but I just don't know how to do it. The ellipse also has 2 radius, an angle and 2 types of feather, I don't know how to adjust all that with the mouse wheel and make it easy for the user.

TurboGit commented:
@edgardoh : I would just show the ellipse and let the mouse wheel adjust the size only. When placed the user can change radius and feather and orientation. I think this would be good enough to help user place the ellipse at the right place with the right size.

edgardoh commented:
ok, if you think that this is good enough I can do it.

TurboGit commented:
@edgardoh : thanks! I really do think this is better than nothing :)

edgardoh commented:
@TurboGit , it's ready, I also added SHIFT + CNTRL to adjust the rotation, to keep it consisten it also works on edit mode.

TurboGit commented:
That's just perfect thanks!

TurboGit commented:
This patch set does not apply on top of retouch branch. There is conflict in develop.h, trivial to fix. Not sure what the best option, rebase this branch on top of retouch or wait for the first branch to be merged and adjust the other. I think this last option is best, nothing to do for now. I just wanted to raise this point.

TurboGit commented:
Would be nice to rebase and fix conflicts on this one. A nice addition... Thanks.

edgardoh commented:
I just rebased and fix conflicts.

This PR hasn't drawn too much attention and it makes some changes on the behaviour of the drawn masks on the blend module.
I find that all the changes here help when working on the retouch, but maybe not everybody like it on the blend module, in particular how masks are resized and the minimum size, so I just want to highlight this new behaviour to make sure everybody is ok with it.

TurboGit commented:
Can you describe what are all the changes in the drawn masks behavior? I'll try to do some more testing soon. Thanks.

edgardoh commented:
-resize don't care where the mouse is inside a shape
Right now, to change the size of a shape, the mouse must be over the "body" of the shape, if it is over the feather, the feather is changed. The feather is also changed when pressing SHIFT, even if the mouse is over the body.
I have changed this so now the size is changed even if the mouse is over the feather, and the feather can only be changed pressing SHIFT.
When there's a lot of small shapes close to each other is difficult to select just the body of a shape, so sometimes I change the feather instead of the body, this change makes the work easer.

Changes in size & step

brush:
I have changed the step from:
    float amount = 1.25f;
    if(up) amount = 0.8f;
to:
      float amount = 1.03f;
      if(up) amount = 0.97f;

and the border from:
        masks_border = MAX(0.005f, MIN(masks_border * amount, 0.5f));
to:
        masks_border = MAX(0.0005f, MIN(masks_border * amount, 0.5f));

this allows to have a smaller brush and more control over the size. For small details I sometimes have to use the path because the brush is not small enough, when the brush will be the right tool for the job.

circle:
I have changed the minimum border size from:
        if(up && circle->border > 0.001f)
to:
        if(up && circle->border > 0.0005f)
this allows to have a harder border.

When the circle is created I have changed the minimum values from:
      circle->radius = MAX(0.01f, circle_size);
      circle->border = MAX(0.005f, circle_border);
to:
      circle->radius = MAX(0.001f, circle_size);
      circle->border = MAX(0.0005f, circle_border);
so the new created circle will have the same values as the last one, even if it is very small, otherwise I'm forced to change the size and feather of every new circle when working with small shapes.

path:
I have changed the minimum border size from:
          masks_border = MAX(0.005f, MIN(masks_border * amount, 0.5f));
to:
          masks_border = MAX(0.0005f, MIN(masks_border * amount, 0.5f));
this allows to have a harder border.

There's also:

-allow to pan & zoom while editing masks
pressing "A" (can be configured) will skip mouse events from the masks, so the image can be "pan&zoom" even if the mouse is over a shape

-preview when creating a circle or ellipse

-resize a path even if on a node or segment
Right now if the mouse is on a path node or segment the mouse wheel will zoom the image, I have fixed that so the path is resized.

TurboGit commented:
Ok, alll this is sounds to me. Even the change of the feather requiring the SHIFT, I had found also that sometimes it was difficult to select just the shape and not the feather.

Do you think we can split the PR in two. I see:
- 1) one commit with the preview for circle & ellipse
- 2) one commit with the behavioral change

This will ease merging, at least I would merge 1 right now. Then will see how 2 looks for others. What do you think?

TurboGit commented on src/develop/masks/ellipse.c:
```diff
@@ -819,6 +1196,7 @@ static void dt_ellipse_events_post_expose(cairo_t *cr, float zoom_scale, dt_mask
 
   float x, y;
 
+  // TODO: this can be replaced with a call to dt_ellipse_draw_shape()
```
and why it is not currently done? maybe add some comment?

edgardoh replied:
because I'm lazy, but I don't want to write that down on the code...
Now seriously, I wanted to make sure it works first and then I just forget, I'll do it on the next commit.

TurboGit replied:
Ok, fine with me. I just wanted to be sure this was not some leftover or forgotten part.

TurboGit commented on src/develop/masks/ellipse.c:
```diff
@@ -881,6 +1259,7 @@ static void dt_ellipse_events_post_expose(cairo_t *cr, float zoom_scale, dt_mask
     }
   }
 
+  // TODO: this can be replaced with a call to dt_ellipse_draw_border()
```
and why it is not currently done? maybe add some comment?

TurboGit commented:
The preview of the shapes are not working on my side with the latest changes. A missing chunk in a commit?

edgardoh commented:
> The preview of the shapes are not working on my side with the latest
> changes. A missing chunk in a commit?
>
Strange, it works fine for me with all the shapes on the blend module, the
spot removal & the retouch.

TurboGit commented:
I have merged it locally on master and it does not work for me :( I had all this working with previous version of the PR. I'll try again...

TurboGit commented:
Still no luck on my side. I have removed my darktable install and build directory. I have merged this PR into my local master (which is exactly origin/master). And I still don't have the preview of the shapes.

TurboGit commented:
BTW, your new PR mask_behavior has conflicts if applied on to of this one. This may indicate that something was broken when you've separated both PR?

TurboGit commented:
Really mysterious! I have built also directly from your branch and it does not work for me :(

I don't have explanation for now. But it doesn't seem to be due to the merge on master.

edgardoh commented:
Are the other features working? The pan&zoom and the path resize with the
mouse over a segment or node...

2018-09-05 12:39 GMT-03:00 Pascal Obry <notifications@github.com>:

> Really mysterious! I have built also directly from your branch and it does
> not work for me :(
>
> I don't have explanation for now. But it doesn't seem to be due to the
> merge on master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/1550#issuecomment-418777724>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APq1pZnhffVmOr6rrThrX2-7SGfP6xuaks5uX_BMgaJpZM4QVYWo>
> .
>

TurboGit commented:
Found... A stupid issue on my side :) I'll check this and merge if all good.

TurboGit commented:
All good and now merged. Thanks.

Please be prepared to write the documentation for the new retouch IOP BTW. You just need to write this as plain text in whatever editor you like. The doc maintainer (Ulrich Pegelow) will ask for this for the next release.

aurelienpierre commented:
@edgardoh Thanks a lot for this beautiful piece of code. It works flawlessly for me, I will test it further in the next days. It's a much needed feature, the result is awesome.

aurelienpierre commented:
@edgardoh I see 2 missing features already:

- we can cut/paste masks from layer to layer, it would be great to be able to reuse the same mask on multiple layers
- the histogram correction in the single layer preview could be automatic so that we don't lose time adjusting white and black levels

I'm still amazed by the abilities of that module though.

TurboGit commented:
Hello Aurélien,

Maybe not what you are looking for but for point 1 above it is possible to use the same masks on a set of layer by using the top slider control. When used, the masks in the "yellow" range of layers are used in all layers. So starting from the first one where it is applied and up where the "noise" is bigger (larger details).

edgardoh commented:
Hi Aurélien,

> @edgardoh <https://github.com/edgardoh> I see 2 missing features already:
>
>    - we can cut/paste masks from layer to layer, it would be great to be
>    able to reuse the same mask on multiple layers
>
> I'm not sure I understand what you are looking for, but if you want to use
all the shapes of a scale on other scales there's the option described by
Pascal above: set the merge from slider and all the edits of a particular
scale will be applied on the range [merge from .. scale where the shape
belongs to] (sorry, I find very difficult to explain what this option does)

>
>    - the histogram correction in the single layer preview could be
>    automatic so that we don't lose time adjusting white and black levels
>
> Yes, is very annoying to set the display wavelet scale button and see a
black image, but I couldn't find a condition that I like. Probably the best
one is:
if in the current session the levels hasn't been used and has default
values, when display wavelet scale is pressed auto levels are executed. But
still I'm not very sure about this.

aurelienpierre commented:
Hello Pascal and Edgardo !

> all the shapes of a scale on other scales there's the option described by Pascal above: set the merge from slider and all the edits of a particular scale will be applied on the range [merge from .. scale where the shape belongs to] (sorry, I find very difficult to explain what this option does)

But then, all the shapes are shared with the selected layers, aren't they ?

> Yes, is very annoying to set the display wavelet scale button and see a
black image, but I couldn't find a condition that I like. Probably the best
one is:

I just would like to map the min intensity of the previewed layer (aka wavelet scale) to 0 and the max to 1.

edgardoh commented:
> But then, all the shapes are shared with the selected layers, aren't they ?
>

The program works like this:

a detail scale is generated
all edits are applied to this scale
the scale is saved somewhere
another scale is generated
all edits are applied to this scale
the scale is merged to the previous scale (both scales already has the
edits applied before they are merge)

if merge from is >= than the scale being generated, then
the scale is generated
it is saved somewhere else
then edits are applied
next scale is generated
it is merged to all previous scales (that has been saved somewhere else)
then edits are applied, so they are applied to all the scales >= merge from
until it reaches the residual image, that it is not affected by this setting

So, in a sense, yes, shapes are shared between wavelet scales (or layers),
but what really happens is that scales are first merged to the previous one
and then the edits are applied.

Maybe you can have a better feeling by just using this feature, the display
wavelet scale option reflect this setting, so what you see in the preview
are the merged scales if current scale > merge from.

I just would like to map the min intensity of the previewed layer (aka
> wavelet scale) to 0 and the max to 1.
>

I have done some experiments with default values for the levels, but it
depends too much on the image. [0, 1] usually don't give good results, at
least on the images I have used, that's why I think that the best solution
(that I can think of) is the one described above.

aurelienpierre commented:
Ok I see. 

For the histogram adjustement on layers, maybe I haven't been clear enough. If you record the min value of the luminance over the desired wavelet scale (called min) and the then max value (called max), you can remap the lumimance channel such as L = (L - min) / (max - min). Then, you will automatically get an output between 0 and 1. You don't need default values, it's adaptable.

Also, is there a mean to rotate the sample patch without rotating the destination patch ? It could be usefull to retouch hair for example.

edgardoh commented:
I think I get it now, but if we are going to do some calculations let's use the auto levels, that seems to work fine for all the images so far. 
But the issue with the condition still remains, When to do the calculations? Each time the display wavelet scale button is pressed? Using the condition I have mentioned?
Maybe instead of wait for the perfect condition I could implement what I already have, if any issues are found I will probably be able to fix it.

@TurboGit , what do you think? Also, is this the right time to introduce this or should we wait after the next release?

About the rotation, no, it can't be done. But this will be a too big change to even try it, we are too close to the next release.

aurelienpierre commented:
> When to do the calculations? Each time the display wavelet scale button is pressed

Yes, as it's just a preview, you could do it anytime for each level, right after the scale computation and just before the displaying on screen. It would make it more readable without having an extra slider. It is a quite standard normalization in signal processing.

As for the rotation, I understand. I wonder if the source code of the cropping module could be used "as is" here.

TurboGit commented:
@edgardoh, it looks sounds to me, I would try Aurélien proposal. And yes it is perfect time to do that now, I have merged well ahead of the release to have field testing & feedback. So better trying to implement good idea before the next release I would say.

edgardoh commented:
I just added PR #1720

aurelienpierre commented:
Thanks !

I have retouched a dozen of pictures now, it works very well, the real problem I have is I don't always have a piece of clean texture in the same orientation than the original texture to patch. I think the sample mask rotation would be the last feature to add to have this module work 100 % of the time.

MRIG commented:
edgardoh,

retouch is not in potfiles.in
did you plan it? will the iop soon be available for translators?

regards Ger

Op wo 5 sep. 2018 om 16:22 schreef edgardoh <notifications@github.com>:

> I just rebased and done a clean build and still works for me. What I've
> done is:
> clone my fork
> checkout retouch_gui branch
> build
>
> When adding a shape on any module, the shape preview is there.
>
> I don't know how the local merge is done, but if you guide me through your
> steps I can try it. Or if you can compare your local files against the
> retouch_gui branch that can give us some info.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/1550#issuecomment-418749313>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB48B1YYNM__vh3qsNUOFnabu6MKQTjwks5uX95CgaJpZM4QVYWo>
> .
>

TurboGit commented:
I have made the French translation, so the translation is present :) Maybe you need to update :+1: 

`$ intltool-update -pot`

?

TurboGit commented:
@edgardoh : I'll have time to review the new PR early next week only.

MRIG commented:
Thanks Pascal,

my mistake, I dit $ intltool-update -pot but forgot to update in poedit
from pot file  ;-(

Op vr 7 sep. 2018 om 21:00 schreef Pascal Obry <notifications@github.com>:

> I have made the French translation, so the translation is present :) Maybe
> you need to update :
>
> $ intltool-update -pot
>
> ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/1550#issuecomment-419533646>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB48Bx4_sNGNkRzkGZ6JTfoyt3IS2Lb5ks5uYsI2gaJpZM4QVYWo>
> .
>


-- 
met vriendelijke groet / kind regards Ger
mailto:g.siemerink@gmail.com
